esphome/components/atm90e32/atm90e32.cpp

```diff
@@ -0,0 +1,192 @@
+  this->write16_(ATM90E32_REGISTER_MMODE0, mmode0);                      // Mode Config (frequency set in main program)
+  this->write16_(ATM90E32_REGISTER_MMODE1, pga_gain_);                   // PGA Gain Configuration for Current Channels
+  this->write16_(ATM90E32_REGISTER_PSTARTTH, 0x0AFC);                    // Active Startup Power Threshold = 50%
+  this->write16_(ATM90E32_REGISTER_QSTARTTH, 0x0AEC);                    // Reactive Startup Power Threshold = 50%
+  this->write16_(ATM90E32_REGISTER_PPHASETH, 0x00BC);                    // Active Phase Threshold = 10%
+  this->write16_(ATM90E32_REGISTER_UGAINA, this->phase_[0].volt_gain_);  // A Voltage rms gain
+  this->write16_(ATM90E32_REGISTER_IGAINA, this->phase_[0].ct_gain_);    // A line current gain
+  this->write16_(ATM90E32_REGISTER_UGAINB, this->phase_[1].volt_gain_);  // B Voltage rms gain
+  this->write16_(ATM90E32_REGISTER_IGAINB, this->phase_[1].ct_gain_);    // B line current gain
+  this->write16_(ATM90E32_REGISTER_UGAINC, this->phase_[2].volt_gain_);  // C Voltage rms gain
+  this->write16_(ATM90E32_REGISTER_IGAINC, this->phase_[2].ct_gain_);    // C line current gain
+  this->write16_(ATM90E32_REGISTER_CFGREGACCEN, 0x0000);                 // end configuration
+}
+
+void ATM90E32Component::dump_config() {
+  ESP_LOGCONFIG("", "ATM90E32:");
+  LOG_PIN("  CS Pin: ", this->cs_);
+  if (this->is_failed()) {
+    ESP_LOGE(TAG, "Communication with ATM90E32 failed!");
+  }
+  LOG_UPDATE_INTERVAL(this);
+  LOG_SENSOR("  ", "Voltage A", this->phase_[0].voltage_sensor_);
+  LOG_SENSOR("  ", "Current A", this->phase_[0].current_sensor_);
+  LOG_SENSOR("  ", "Power A", this->phase_[0].power_sensor_);
+  LOG_SENSOR("  ", "Voltage B", this->phase_[1].voltage_sensor_);
+  LOG_SENSOR("  ", "Current B", this->phase_[1].current_sensor_);
+  LOG_SENSOR("  ", "Power B", this->phase_[1].power_sensor_);
+  LOG_SENSOR("  ", "Voltage C", this->phase_[2].voltage_sensor_);
+  LOG_SENSOR("  ", "Current C", this->phase_[2].current_sensor_);
+  LOG_SENSOR("  ", "Power C", this->phase_[2].power_sensor_);
+  LOG_SENSOR("  ", "Frequency", this->freq_sensor_)
+}
+float ATM90E32Component::get_setup_priority() const { return setup_priority::DATA; }
+
+uint16_t ATM90E32Component::read16_(uint16_t a_register) {
+  uint8_t addrh = (1 << 7) | ((a_register >> 8) & 0x03);
+  uint8_t addrl = (a_register & 0xFF);
+  uint8_t data[2];
+  uint16_t output;
+
+  this->enable();
+  delayMicroseconds(10);
+  this->write_byte(addrh);
+  this->write_byte(addrl);
+  delayMicroseconds(4);
+  this->read_array(data, 2);
+  this->disable();
+
+  output = (uint16_t(data[0] & 0xFF) << 8) | (data[1] & 0xFF);
+  ESP_LOGVV(TAG, "read16_ 0x%04X output 0x%04X", a_register, output);
+  return output;
+}
+
+int ATM90E32Component::read32_(uint16_t addr_h, uint16_t addr_l) {
+  int val_h = this->read16_(addr_h);
+  int val_l = this->read16_(addr_l);
+  int val = (val_h << 16) | val_l;
```

Comment:
```suggestion
  uint16_t val_h = this->read16_(addr_h);
  uint16_t val_l = this->read16_(addr_l);
  int32_t val = (val_h << 16) | val_l;
```

Explicitly type the values here - the previous code worked too but this makes it more portable.